packages/editor/core/src/ui/plugins/image/image-upload-handler.ts

```diff
@@ -0,0 +1,112 @@
+import { type UploadImage } from "src/types/upload-image";
+
+// utilities
+import { v4 as uuidv4 } from "uuid";
+
+// types
+import { isFileValid } from "src/ui/plugins/image/utils/validate-file";
+import { Editor } from "@tiptap/core";
+import { EditorView } from "@tiptap/pm/view";
+import { uploadKey } from "./constants";
+import { removePlaceholder, findPlaceholder } from "./utils/placeholder";
+
+export async function startImageUpload(
+  editor: Editor,
+  file: File,
+  view: EditorView,
+  pos: number | null,
+  uploadFile: UploadImage
+) {
+  editor.storage.image.uploadInProgress = true;
+
+  if (!isFileValid(file)) {
+    editor.storage.image.uploadInProgress = false;
+    return;
+  }
+
+  const id = uuidv4();
+
+  const tr = view.state.tr;
+  if (!tr.selection.empty) tr.deleteSelection();
+
+  const reader = new FileReader();
+  reader.readAsDataURL(file);
+  reader.onload = () => {
+    tr.setMeta(uploadKey, {
+      add: {
+        id,
+        pos,
+        src: reader.result,
+      },
+    });
+    view.dispatch(tr);
+  };
+
+  // Handle FileReader errors
+  reader.onerror = (error) => {
+    console.error("FileReader error: ", error);
+    removePlaceholder(editor, view, id);
+    return;
+  };
+
+  try {
+    const src = await uploadAndValidateImage(file, uploadFile);
+
+    if (src == null) {
+      throw new Error("Resolved image URL is undefined.");
+    }
+
+    const { schema } = view.state;
+    pos = findPlaceholder(view.state, id);
+
+    if (pos == null) {
+      editor.storage.image.uploadInProgress = false;
```

Comment:
Shouldn't this be outside the if condition?

Reply:
It was actually meant to set the progress as false if the position is null (while trying to find the placeholder) due to any reason. Please let me know if there's a better way I could go about this, thanks 😄